File: Python/Arrays and Hashing/product_array_except_self.py
# 238. Product of Array Except Self
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = []

        prefix = []
        # Calculate all the prefix
        prefix_aux = 1
        for i in range(len(nums)):
            prefix.append(prefix_aux)
            prefix_aux *= nums[i]

        sufix = (
            [0] * len(nums)
        )  # initialize an array with 0 -> important cause we need to fill the array since the last element
        # Calculate all the sufix
        sufix_aux = 1
        for i in range(
            len(nums) - 1, -1, -1
        ):  # the las two -1 means -> repeat until the -1th element and decrece in 1 each step
            sufix[i] = sufix_aux
            sufix_aux *= nums[i]

        for i in range(len(nums)):
            res.append(prefix[i] * sufix[i])

        return res

File: cpp/Arrays and Hashing/ProductArrayExceptSelf.cpp
// 238. Product of Array Except Self

#include <vector>
#include <iostream>
#include <set>
#include <unordered_set>
using namespace std;

class Solution {
    public:
        vector<int> productExceptSelf(vector<int>& nums) {
            vector<int> res;    
            int len = nums.size();        

            // calculate the list of prefix
            vector<int> prefix;
            int aux_prefix = 1;
            for (int i = 0; i < len; i++){
                prefix.push_back(aux_prefix);
                aux_prefix *= nums[i];
            }
            // calculate the list of sufix
            
            vector<int> sufix(len,0);   // fixed size and initialize in 0
            int aux_sufix = 1;
            for(int i = len - 1; i >= 0 ; i--){ // until 0 cause we neet to iterate over nums[0] lol
                sufix[i] = aux_sufix;
                aux_sufix *= nums[i];
            }

            for(int i = 0; i < len; i++){
                res.push_back(prefix[i] * sufix[i]);
            }
            return res;
        }
};